File: test/language/templates/python/SubtypeTemplateWithCompoundTest.py
import zserio

import Templates

class SubtypeTemplateWithCompoundTest(Templates.TestCase):
    def testReadWrite(self):
        subtypeTemplateWithCompound = self.api.SubtypeTemplateWithCompound(
            self.api.Compound(13),
            self.api.TemplateCompound_Compound(self.api.Compound(42))
        )

        writer = zserio.BitStreamWriter()
        subtypeTemplateWithCompound.write(writer)
        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        readSubtypeTemplateWithCompound = self.api.SubtypeTemplateWithCompound()
        readSubtypeTemplateWithCompound.read(reader)
        self.assertEqual(subtypeTemplateWithCompound, readSubtypeTemplateWithCompound)

File: test/language/templates/cpp/SubtypeTemplateWithCompoundTest.cpp
#include "gtest/gtest.h"
#include "templates/subtype_template_with_compound/SubtypeTemplateWithCompound.h"

namespace templates
{
namespace subtype_template_with_compound
{

TEST(SubtypeTemplateWithCompoundTest, readWrite)
{
    SubtypeTemplateWithCompound subtypeTemplateWithCompound;
    subtypeTemplateWithCompound.setValue1(Compound{13});
    subtypeTemplateWithCompound.setValue2(TemplateCompound_Compound{Compound{42}});

    zserio::BitBuffer bitBuffer = zserio::BitBuffer(1024 * 8);
    zserio::BitStreamWriter writer(bitBuffer);
    subtypeTemplateWithCompound.write(writer);

    zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
    SubtypeTemplateWithCompound readSubtypeTemplateWithCompound(reader);

    ASSERT_TRUE(subtypeTemplateWithCompound == readSubtypeTemplateWithCompound);
}

} // namespace subtype_template_with_compound
} // namespace templates
